modules/directives/src/main/scala/scala/build/preprocessing/directives/UsingSourceDirectiveHandler.scala

```diff
@@ -0,0 +1,56 @@
+package scala.build.preprocessing.directives
+
+import scala.build.Logger
+import scala.build.Ops._
+import scala.build.errors.{BuildException, CompositeBuildException, WrongSourcePathError}
+import scala.build.options.{BuildOptions, ClassPathOptions}
+import scala.util.{Failure, Success}
+
+case object UsingSourceDirectiveHandler extends UsingDirectiveHandler {
+  def name        = "Custom sources"
+  def description = "Manually add sources to the Scala CLI project"
+  def usage       = "`//> using file `_path_ | `//> using files `_path1_, _path2_ …"
+  override def usageMd =
+    """//> using file hello.sc
+      |
+      |//> using files Utils.scala, Helper.scala …""".stripMargin
+
+  override def examples = Seq(
+    "//> using file \"utils.scala\""
+  )
+
+  def keys = Seq("file", "files")
+  def handleValues(
+    scopedDirective: ScopedDirective,
+    logger: Logger
+  ): Either[BuildException, ProcessedUsingDirective] =
+    checkIfValuesAreExpected(scopedDirective).flatMap { groupedScopedValuesContainer =>
+      groupedScopedValuesContainer.scopedStringValues.map {
+        case ScopedValue(positioned, _) =>
+          val eitherRootPathOrBuildException =
+            Directive.osRoot(scopedDirective.cwd, positioned.positions.headOption)
+          eitherRootPathOrBuildException.flatMap { root =>
+            scala.util.Try(os.Path(positioned.value, root)) match {
+              case Failure(exception) =>
+                Left(new WrongSourcePathError(exception.getLocalizedMessage))
+              case Success(sourcePath) => Right(sourcePath)
+            }
+          }
+      }
+        .sequence
+        .left.map(CompositeBuildException(_))
+
+    }.map { pathSequence =>
+
+      ProcessedDirective(
+        Some(BuildOptions(
+          classPathOptions = ClassPathOptions(
+            extraSourceFiles = pathSequence
+          )
+        )),
+        Seq.empty
+      )
+
+    }
```

Comment:
NIT (readability)
```suggestion
    checkIfValuesAreExpected(scopedDirective)
      .flatMap { groupedScopedValuesContainer =>
        groupedScopedValuesContainer.scopedStringValues
          .map {
            case ScopedValue(positioned, _) =>
              val eitherRootPathOrBuildException =
                Directive.osRoot(scopedDirective.cwd, positioned.positions.headOption)
              eitherRootPathOrBuildException.flatMap { root =>
                Try(os.Path(positioned.value, root)).toEither.left.map { ex =>
                  new WrongSourcePathError(exception.getLocalizedMessage)
                }
              }
          }
          .sequence
          .left.map(CompositeBuildException(_))

      }
      .map { pathSequence =>
        ProcessedDirective(
          Some(BuildOptions(
            classPathOptions = ClassPathOptions(
              extraSourceFiles = pathSequence
            )
          )),
          Seq.empty
        )
      }
```

Reply:
Or even, with `either` / `value`
```suggestion
    val groupedScopedValuesContainer = value(checkIfValuesAreExpected(scopedDirective))
    val pathSequence = value {
      groupedScopedValuesContainer.scopedStringValues
        .map {
          case ScopedValue(positioned, _) =>
            val eitherRootPathOrBuildException =
              Directive.osRoot(scopedDirective.cwd, positioned.positions.headOption)
            eitherRootPathOrBuildException.flatMap { root =>
              Try(os.Path(positioned.value, root)).toEither.left.map { ex =>
                new WrongSourcePathError(ex.getLocalizedMessage)
              }
            }
        }
        .sequence
        .left.map(CompositeBuildException(_))
    }
      
      ProcessedDirective(
        Some(BuildOptions(
          classPathOptions = ClassPathOptions(
            extraSourceFiles = pathSequence
          )
        )),
        Seq.empty
      )
```

Reply:
We can also keep the cause of the underlying exception `ex` in `WrongSourcePathError` (just pass `cause` in its constructor to `BuildException`), that can help sometimes.